src/query_plan/query_planning_traversal.rs

```diff
@@ -86,3 +89,108 @@ struct BestQueryPlanInfo {
     /// The cost of this query plan.
     cost: QueryPlanCost,
 }
+
+impl QueryPlanningTraversal {
+    pub(crate) fn new_dependency_graph(&self) -> FetchDependencyGraph {
+        let root_type = if self.is_top_level && self.has_defers {
+            self.parameters
+                .supergraph_schema
+                .schema()
+                .root_operation(self.root_kind.into())
+                .cloned()
+                // A root operation type has to be an object type
+                .map(|type_name| ObjectTypeDefinitionPosition { type_name }.into())
+        } else {
+            None
+        };
+        FetchDependencyGraph::new(
+            self.parameters.supergraph_schema.clone(),
+            self.parameters.federated_query_graph.clone(),
+            root_type,
+            self.starting_id_generation,
+        )
+    }
+
+    fn updated_dependency_graph(
+        &self,
+        dependency_graph: &mut FetchDependencyGraph,
+        path_tree: &OpPathTree,
+    ) -> Result<(), FederationError> {
+        let is_root_path_tree = matches!(
+            path_tree.graph.node_weight(path_tree.node)?.type_,
+            QueryGraphNodeType::SchemaType(_)
```

Comment:
It's inverted here, the path tree is only a root path tree if the `type_` matches `QueryGraphNodeType::FederatedRootType(_)` instead of `SchemaType`.